File: music_car/android/app/src/main/java/com/icar/musiccar/network/ApiClient.kt
/**
 * 网络层 — HTTP 客户端（Android / Kotlin）
 *
 * 对应鸿蒙端的 ApiClient.ets，严格复用 APP接口文档.md 的 HTTP 接口。
 *
 * 依赖: OkHttp + kotlinx.serialization
 *
 * 接口：
 *   POST /api/cmd              — 控制指令
 *   POST /api/process/start    — 启动功能节点
 *   POST /api/process/stop     — 停止功能节点
 *   GET  /api/state            — 获取状态（一次性快照）
 */

package com.icar.musiccar.network

import com.icar.musiccar.AppConstants
import com.icar.musiccar.models.CmdRequest
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.encodeToString
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

// ── 数据模型 ──────────────────────────────────────

@Serializable
data class ApiResponse(
    val ok: Boolean,
    val error: String? = null,
    val pid: Int? = null
)

@Serializable
data class CarState(
    val linear_x: Double,
    val angular_z: Double,
    val estop: Boolean,
    val mode: String,
    val running_nodes: List<String> = emptyList()
)

@Serializable
data class ProcessRequest(
    val function: String
)

// ── HTTP 客户端 ───────────────────────────────────

class ApiClient {
    companion object {
        @Volatile
        private var instance: ApiClient? = null

        fun getInstance(): ApiClient {
            return instance ?: synchronized(this) {
                instance ?: ApiClient().also { instance = it }
            }
        }
    }

    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private val client: OkHttpClient
    private var baseUrl: String = "http://${AppConstants.DEFAULT_CAR_IP}:${AppConstants.DEFAULT_CAR_PORT}"
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    init {
        client = OkHttpClient.Builder()
            .connectTimeout(AppConstants.HTTP_TIMEOUT, TimeUnit.MILLISECONDS)
            .readTimeout(AppConstants.HTTP_TIMEOUT, TimeUnit.MILLISECONDS)
            .build()
    }

    /** 设置小车地址 */
    fun setServer(ip: String, port: Int = 5000) {
        baseUrl = "http://$ip:$port"
    }

    // ── 公开接口 ────────────────────────────────

    /** 发送控制指令 */
    suspend fun postCmd(cmd: CmdRequest): ApiResponse {
        return post("/api/cmd", cmd)
    }

    /** 启动功能节点 */
    suspend fun startProcess(funcName: String): ApiResponse {
        return post("/api/process/start", ProcessRequest(funcName))
    }

    /** 停止功能节点 */
    suspend fun stopProcess(funcName: String): ApiResponse {
        return post("/api/process/stop", ProcessRequest(funcName))
    }

    /** 获取小车状态（一次性快照） */
    suspend fun getState(): CarState {
        val request = Request.Builder()
            .url("$baseUrl/api/state")
            .get()
            .build()

        return withIO {
            val response = client.newCall(request).execute()
            if (response.isSuccessful) {
                val body = response.body?.string() ?: throw IOException("空响应")
                json.decodeFromString<CarState>(body)
            } else {
                throw IOException("HTTP ${response.code}")
            }
        }
    }

    /** 测试连通性 */
    suspend fun ping(): Boolean {
        return try {
            getState()
            true
        } catch (_: Exception) {
            false
        }
    }

    // ── 内部方法 ────────────────────────────────

    private suspend fun post(path: String, data: Any): ApiResponse {
        val body = json.encodeToString(data).toRequestBody(jsonMediaType)
        val request = Request.Builder()
            .url("$baseUrl$path")
            .post(body)
            .build()

        return withIO {
            try {
                val response = client.newCall(request).execute()
                if (response.isSuccessful) {
                    val bodyStr = response.body?.string() ?: "{}"
                    json.decodeFromString<ApiResponse>(bodyStr)
                } else {
                    ApiResponse(ok = false, error = "HTTP ${response.code}")
                }
            } catch (e: IOException) {
                ApiResponse(ok = false, error = "网络错误: ${e.message}")
            }
        }
    }

    private suspend fun <T> withIO(block: () -> T): T {
        return kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
            block()
        }
    }
}

File: music_car/ios/MusicCar/Network/ApiClient.swift
/**
 * 网络层 — HTTP 客户端（iOS / Swift）
 *
 * 对应鸿蒙端的 ApiClient.ets，严格复用 APP接口文档.md 的 HTTP 接口。
 *
 * 接口：
 *   POST /api/cmd              — 控制指令
 *   POST /api/process/start    — 启动功能节点
 *   POST /api/process/stop     — 停止功能节点
 *   GET  /api/state            — 获取状态（一次性快照）
 */

import Foundation

class ApiClient {
    static let shared = ApiClient()
    
    private var baseURL: String = "http://\(AppConstants.defaultCarIP):\(AppConstants.defaultCarPort)"
    private let session: URLSession
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()
    
    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = AppConstants.httpTimeout
        config.timeoutIntervalForResource = AppConstants.httpTimeout
        session = URLSession(configuration: config)
    }
    
    /// 设置小车地址
    func setServer(ip: String, port: Int = 5000) {
        baseURL = "http://\(ip):\(port)"
    }
    
    // ── 公开接口 ────────────────────────────────
    
    /// 发送控制指令
    func postCmd(_ cmd: CmdRequest) async throws -> ApiResponse {
        return try await post(path: "/api/cmd", body: cmd)
    }
    
    /// 启动功能节点
    func startProcess(_ funcName: String) async throws -> ApiResponse {
        return try await postDict(path: "/api/process/start", dict: ["function": funcName])
    }
    
    /// 停止功能节点
    func stopProcess(_ funcName: String) async throws -> ApiResponse {
        return try await postDict(path: "/api/process/stop", dict: ["function": funcName])
    }
    
    /// 获取小车状态（一次性快照）
    func getState() async throws -> CarState {
        guard let url = URL(string: "\(baseURL)/api/state") else {
            throw URLError(.badURL)
        }
        
        let (data, response) = try await session.data(from: url)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw URLError(.badServerResponse)
        }
        
        return try decoder.decode(CarState.self, from: data)
    }
    
    /// 测试连通性
    func ping() async -> Bool {
        do {
            _ = try await getState()
            return true
        } catch {
            return false
        }
    }
    
    // ── 内部方法 ────────────────────────────────
    
    private func post<T: Encodable>(path: String, body: T) async throws -> ApiResponse {
        guard let url = URL(string: "\(baseURL)\(path)") else {
            throw URLError(.badURL)
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try encoder.encode(body)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        
        if httpResponse.statusCode == 200 {
            return try decoder.decode(ApiResponse.self, from: data)
        }
        
        return ApiResponse(ok: false, error: "HTTP \(httpResponse.statusCode)")
    }

    /// 字典版 POST（用于简单键值对请求，如 process/start）
    private func postDict(path: String, dict: [String: String]) async throws -> ApiResponse {
        guard let url = URL(string: "\(baseURL)\(path)") else {
            throw URLError(.badURL)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: dict)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            return ApiResponse(ok: false, error: "HTTP \((response as? HTTPURLResponse)?.statusCode ?? 0)")
        }

        return try decoder.decode(ApiResponse.self, from: data)
    }
}
